### ISSUE TITLE
 hashCode for doubles should be better.

### ISSUE THREAD

**Issue by DartBot** [dart-lang/sdk#2884]
_This issue was originally filed by @mdakin_

---

**What steps will reproduce the problem?**
1. Run http://try.dartlang.org/s/P6E_
2. It seems hashCode for double numbers is only the integer part of the number.so hashCode of 0.1 and 0.9 are both 0.

**What is the expected output? What do you see instead?**
hashCode for different double numbers should be different.

**Comment by sethladd** [comment]
This is still the case.

```
main() {
  print(0.2345.hashCode);  // 0
  print(1.9999.hashCode);  // 1
  print(1.00000.hashCode);  // 1
  print(123.456.hashCode);  // 123
}
```

**Comment by Hixie** [comment]
This seems like it might cause problems in Flutter.
Many of our objects have hashCode getters that derive their hashCode from the hashCodes of their properties. Many of our objects have double properties. Many of them in particular have a double in the range 0.0..1.0, with no other changes.

For example, if you use FlutterLogoDecoration.lerp to fade in a logo, you'll get back a series of objects that are identical except for their opacity, which will vary from 0.0 to 1.0. Say you want to fade something in over 10 seconds. If you stick all of these objects in a Map for efficiency, you'll get hash collisions on all values except the last one, so you'll have one bucket with 599 objects and one bucket with 1 object.

**Comment by kasperl** [comment]
Let me take a stab at this!

**Comment by kasperl** [comment]
Started working on this: https://dart-review.googlesource.com/c/sdk/+/15140. 

```dart
main() {
  print(0.2345.hashCode);  // 4597616773615499378
  print(1.9999.hashCode);  // 4611685567917960216
  print(1.00000.hashCode);  // 1
  print(123.456.hashCode);  // 26701843266626392
}
```

**Comment by Hixie** [comment]
I am so curious to see what impact this has on our benchmarks!

**Comment by kasperl** [comment]
Fixed in 1db8615e29cf1e62854d041b805c6d3c43c2225c.
